pkg/pool-weighted/contracts/lib/CircuitBreakerLib.sol

```diff
@@ -0,0 +1,299 @@
+        // the circuit breaker for the token in that direction.
+        _require(params.lowerBound <= FixedPoint.ONE, Errors.INVALID_CIRCUIT_BREAKER_BOUNDS);
+        _require(params.upperBound <= _MAX_BOUND_PERCENTAGE, Errors.INVALID_CIRCUIT_BREAKER_BOUNDS);
+        _require(
+            params.upperBound == 0 || params.upperBound >= params.lowerBound,
+            Errors.INVALID_CIRCUIT_BREAKER_BOUNDS
+        );
+
+        // Set the reference parameters: BPT price of the token, and the weight complement.
+        bytes32 circuitBreakerState = bytes32(0)
+            .insertUint(params.bptPrice, _BPT_PRICE_OFFSET, _BPT_PRICE_WIDTH)
+            .insertUint(params.weightComplement, _WEIGHT_COMPLEMENT_OFFSET, _WEIGHT_COMPLEMENT_WIDTH);
+
+        // Add the lower and upper percentage bounds.
+        circuitBreakerState = circuitBreakerState
+            .insertUint(
+            params.lowerBound.compress(_BOUND_WIDTH, _MAX_BOUND_PERCENTAGE),
+            _LOWER_BOUND_OFFSET,
+            _BOUND_WIDTH
+        )
+            .insertUint(
+            params.upperBound.compress(_BOUND_WIDTH, _MAX_BOUND_PERCENTAGE),
+            _UPPER_BOUND_OFFSET,
+            _BOUND_WIDTH
+        );
+
+        // Precompute and store the conversion ratios, used to convert percentage bounds to BPT price bounds.
+        // If the weight complement has not changed since the breaker was set (i.e., if there is no ongoing weight
+        // update, and no tokens have been added or removed), we can use the precomputed values directly, and avoid
+        // a heavy computation.
+        (uint256 lowerBoundRatio, uint256 upperBoundRatio) = getBoundaryConversionRatios(
+            params.lowerBound,
+            params.upperBound,
+            params.weightComplement
+        );
+
+        // Finally, insert these computed ratios, and return the complete set of fields.
+        return
+            circuitBreakerState
+                .insertUint(
+                lowerBoundRatio.compress(_BOUND_RATIO_WIDTH, _MAX_BOUND_PERCENTAGE),
+                _LOWER_BOUND_RATIO_OFFSET,
+                _BOUND_RATIO_WIDTH
+            )
+                .insertUint(
+                upperBoundRatio.compress(_BOUND_RATIO_WIDTH, _MAX_BOUND_PERCENTAGE),
+                _UPPER_BOUND_RATIO_OFFSET,
+                _BOUND_RATIO_WIDTH
+            );
+    }
+
+    // Convert percentage bounds to BPT price bounds
+    function getBoundaryConversionRatios(
+        uint256 lowerBound,
+        uint256 upperBound,
+        uint256 currentWeightComplement
+    ) internal pure returns (uint256 lowerBoundRatio, uint256 upperBoundRatio) {
+        // Rounding down for the lower bound, and up for the upper bound will maximize the
+        // "operating range" - the BPT price range that will not trigger the circuit breaker -
+        // of the pool for traders.
+        lowerBoundRatio = lowerBound.powDown(currentWeightComplement);
```

Comment:
Related to https://github.com/balancer-labs/balancer-v2-monorepo/issues/1844:

If I recall correctly, for `pow(x,y)` to be valid, then `log(x) * y > -41`. The worst case for `y` is 0.99, which seems to mean that `x` could indeed go extremely low. Do we need to though? Why can the lower bound be so small but the upper bound is just 10x?

Reply:
This all needs to be revisited given the switch to normalized weights. I updated the code, but the test range can now be expanded.